### ISSUE TITLE
Grouping sets fails on struct columns with dot notation

### ISSUE THREAD

**Issue by cploonker** [trinodb/trino#22604]
# Following query works
```
SELECT pr.name as name, MAX(pr.age) as age, MAX(city) as city
FROM (
  SELECT CAST(pr AS ROW(name VARCHAR, age int)) as pr, city
  FROM(
    VALUES 
      (ROW('John Doe', 30), 'New York'),
      (ROW('Jane Doe', 25), 'Los Angeles')
  ) AS t(pr, city)
)
GROUP BY pr.name
```

# However following query fails
Shouldn't both work?
```
SELECT pr.name as name, MAX(pr.age) as age, MAX(city) as city
FROM (
  SELECT CAST(pr AS ROW(name VARCHAR, age int)) as pr, city
  FROM(
    VALUES 
      (ROW('John Doe', 30), 'New York'),
      (ROW('Jane Doe', 25), 'Los Angeles')
  ) AS t(pr, city)
)
GROUP BY GROUPING SETS (pr.name)
```

**Comment by cploonker** [comment]
Presto has the same issue: https://github.com/prestodb/presto/issues/23151

**Comment by hashhar** [comment]
cc: @martint

**Comment by martint** [comment]
Per the SQL specification:

> Each `<grouping column reference>` shall unambiguously reference a column of the table resulting from the `<from clause>`

The way the `GROUPING SETS`, `ROLLUP` and `CUBE` clauses are implemented follow the specification to the letter. Grouping by an expression (in the case of a vanilla `GROUP BY`) is an extension to the specification, which predates the implementation of those more advance grouping set qualifiers. We could consider adding the same extensions to those constructs, but there are some non-trivial implications for how [`grouping()`](https://trino.io/docs/current/sql/select.html#grouping-operation) and other operations such as non-deterministic functions would work, and that would need to be sorted out.

**Comment by cploonker** [comment]
Is a struct field not considered a column?

**Comment by martint** [comment]
No, that’s considered a “complex expression” (a field dereference expression)

**Comment by cploonker** [comment]
Got it. Thanks for the explaination.
So this is a case of SQL standard coming in the way of ease of use.

**Comment by bristy** [comment]
@cploonker  Is there any workaround for this limitation?
